vowpalwabbit/conditional_contextual_bandit.cc

```diff
@@ -1,23 +1,147 @@
 #include "conditional_contextual_bandit.h"
-
+#include "reductions.h"
 #include "example.h"
 #include "global_data.h"
 #include "cache.h"
 
 #include <numeric>
 #include <algorithm>
 
+using namespace LEARNER;
 using namespace VW;
+using namespace VW::config;
+
+template <bool is_learn>
+void do_actual_learning(CCB::ccb& data, multi_learner& base, multi_ex& examples) {
+
+  //get decisions, actions and shared parts of the multiline example
+  std::vector<example*> decisions, actions;
+  example* shared;
+  for (auto ex : examples) {
+    switch (ex->l.conditional_contextual_bandit.type) {
+      case CCB::example_type::shared:
+        shared = ex;
+        break;
+      case CCB::example_type::action:
+        actions.push_back(ex);
+        break;
+      case CCB::example_type::decision:
+        decisions.push_back(ex);
+        break;
+      default:
+        THROW("ccb_adf_explore: badly formatted example - invalid example type");
+    }
+  }
 
-bool CCB::ec_is_example_header(example& ec)
-{
-  return ec.l.conditional_contextual_bandit.type == example_type::shared;
+  //sanity check
+  if (decisions.size() > actions.size())
+    THROW("ccb_adf_explore: badly formatted example - number of actions " << actions.size() << " must be greater than the number of decisions " << decisions.size());
+
+  //TODO inject decisions into shared (share := shared + decision features)
+  //TODO add tests and UT
+
+  //set the CB label in the shared example
+  //set the proba to -1 and the cost to float_max (convention for CB to identify shared data)
+  CB::cb_class f;
+  f.probability = -1.f;
+  f.cost = FLT_MAX;
+  shared->l.cb.costs = v_init<CB::cb_class>();
+  shared->l.cb.costs.push_back(f);
+
+  std::set<uint32_t> blacklist_actions, whitelist_actions;
+  CCB::decision_scores_t decision_scores = v_init< ACTION_SCORE::action_scores>();
```

Comment:
This object should deleted when the prediction type is deleted.
todo: override the delete function